pkg/volume/csi/csi_attacher.go

```diff
@@ -458,27 +416,46 @@ func (c *csiAttacher) waitForVolumeDetachment(volumeHandle, attachID string) err
 	return c.waitForVolumeDetachmentInternal(volumeHandle, attachID, timer, timeout)
 }
 
-func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID string, timer *time.Timer, timeout time.Duration) error {
+func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID string, timer *time.Timer,
+	timeout time.Duration) (string, error) {
 	klog.V(4).Info(log("probing VolumeAttachment [id=%v]", attachID))
 	attach, err := c.k8s.StorageV1().VolumeAttachments().Get(attachID, meta.GetOptions{})
 	if err != nil {
 		if apierrs.IsNotFound(err) {
 			//object deleted or never existed, done
 			klog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] not found, object deleted", attachID, volumeHandle))
-			return nil
+			return attachID, nil
 		}
-		return errors.New(log("detacher.WaitForDetach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
+		klog.Error(log("detacher.WaitForDetach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
+		return "", err
 	}
-	// driver reports attach error
-	detachErr := attach.Status.DetachError
+
+	return c.waitForVolumeAttachDetachmentStatus(attach, volumeHandle, attachID, timer, timeout, verifyDetachmentStatus, c.waitForVolumeDetachmentInternal)
+}
+
+func verifyDetachmentStatus(attachment *storage.VolumeAttachment, volumeHandle string) (bool, error) {
+	// driver reports detach error
+	detachErr := attachment.Status.DetachError
 	if detachErr != nil {
-		klog.Error(log("detachment for VolumeAttachment [%v] for volume [%s] failed: %v", attachID, volumeHandle, detachErr.Message))
-		return errors.New(detachErr.Message)
+		klog.Error(log("detachment for VolumeAttachment for volume [%s] failed: %v", volumeHandle, detachErr.Message))
+		return false, errors.New(detachErr.Message)
+	}
+	return false, nil
+}
+
+func (c *csiAttacher) waitForVolumeAttachDetachmentStatus(attach *storage.VolumeAttachment, volumeHandle, attachID string,
+	timer *time.Timer, timeout time.Duration, waitForStatus waitForStatus, waitForReadyInternal waitForStatusInternal) (string, error) {
```

Comment:
Only `attachID` is returned as the `string` return value. Can this function just return error and caller can use `attachID` when `waitForVolumeAttachDetachmentStatus` returns success?